mypy/newsemanal/semanal.py

```diff
@@ -3989,10 +3989,10 @@ def lookup_fully_qualified(self, name: str) -> SymbolTableNode:
         module namespace is ignored.
         """
         parts = name.split('.')
-        n = self.modules[parts[0]]
+        n = self.modules[parts[0]]  # type: Union[MypyFile, TypeInfo]
         for i in range(1, len(parts) - 1):
             next_sym = n.names[parts[i]]
-            assert isinstance(next_sym.node, MypyFile)
+            assert isinstance(next_sym.node, (MypyFile, TypeInfo))
```

Comment:
I think there is some misunderstanding/bug somewhere. This function should never be used for things like local lookup. It should be only used to construct a known type like `builtins.str`. For local lookup one needs to use `lookup_qualified()`. Why do you need this modification?

Reply:
I made this change in order to get this example working:

    from enum import Enum
    from typing_extensions import Literal

    class Outer:
        class Inner(Enum):
            FOO = 1
            BAR = 2

    x: Literal[Outer.Inner.FOO]
    reveal_type(x)

Basically, the literal-handling code in `typeanal` receives back the string "module.Outer.Inner". We then call `typeanal.named_type_with_normalized_str` which in turn calls `typeanal.named_type` which in turn calls `semanal.lookup_fully_qualified`. That function then ends up crashing because "Outer" is a TypeInfo, not a MypyFile.

FWIW I was under the impression that:

1. `lookup_qualified` is for looking up fully-qualified types, and has a bunch of error handling for when the thing we're trying to look up doesn't exist.
2. `lookup_fully_qualified` is a simpler version of `lookup_qualified`, except without error handling.
3. `lookup` is for looking up some unqualified name across all namespaces (local and global).

I also noticed that `lookup_qualified` handles both MypyFile and TypeInfo, and so figured that the fact that `lookup_fully_qualified` doesn't was an oversight.

Reply:
> FWIW I was under the impression that

This is a wrong impression. Which is normal, we have more than dozen lookup functions in the code-base (likely caused by the fact that many people who are not familiar with mypy contributed some reinvented bicycles, which is again normal). Largely, they can be separated in two classes:
* Local lookup: what given name or member expression points to using Python name resolution
* Global lookup: find a symbol using its "full address" used to construct some standard known types.

The `named_type()` and friends are from the second group. While the first group is roughly two methods `lookup()` and `lookup_qualified()`.

Essentially what is going on here looks like a hack (and TBH I don't like it). You already have the `TypeInfo` needed to construct the literal type, but instead you get its full name and pass it to `RawExpressionType` that later passes the name to find the same `TypeInfo` again.

Initially, string name was passed to `RawLiteralExpression` because it is was constructed at very early stage. Here however, it looks unnecessary. Why can't we just immediately return `LiteralType` there with a fallback instance constructed from a known `TypeInfo`?

Reply:
> The `named_type()` and friends are from the second group. While the first group is roughly two methods `lookup()` and `lookup_qualified()`.

Hmm, ok. In that case, what's the correct way of looking up a nested type via a global lookup?

It seems all of the "global lookup" functions assume that all types are defined within modules only, which I'm not so sure is correct.

> Why can't we just immediately return `LiteralType` there with a fallback instance constructed from a known `TypeInfo`?

Sure, we can do that. Is it safe to just construct new whenever we want Instances though? E.g. there isn't a need to register them in some global registry?

Reply:
> It seems all of the "global lookup" functions assume that all types are defined within modules only, which I'm not so sure is correct.

It is not correct, but this is kind of optimization. After the refactoring we will have two methods, one fast for builtin things like `builtins.str`, and another one that can be used in plugins etc., with full semantics.

> E.g. there isn't a need to register them in some global registry?

What kind of registry do you mean? I think one can just call `Instance(info, args, line, column)` or am I missing something?.

Reply:
> What kind of registry do you mean? I think one can just call Instance(info, args, line, column) or am I missing something?.

Honestly, I have no idea. Instances and TypeInfos are stuffed with so much functionality that I always get a little worried that touching either of them or creating new ones will break something or interfere with something related to incremental or fine-grained mode in some weird way.